### ISSUE TITLE
[23] Folding a markdown region hides the first line of subsequent Java code

### ISSUE THREAD

**Issue by stephan-herrmann** [eclipse-jdt/eclipse.jdt.ui#1615]
Given 
```
	///
	/// This is the comment of m()
	///
	void m() {
		System.out.println("m");
	}
```
when you fold the markdown comment in the editor, it looks like this:

![image](https://github.com/user-attachments/assets/f95eb02c-951d-45de-980c-ebfa1465a55f)

@jarthana looks like your latest change in #1549 only moved the problem :)

**Comment by stephan-herrmann** [comment]
@jarthana  have you seen this?

**Comment by jarthana** [comment]
> @jarthana have you seen this?

Thanks, will take a look.

**Comment by jarthana** [comment]
Actually, I don't see fold/unfold option at all for markdown comments. Sigh! Let me see what's going on.

**Comment by jarthana** [comment]
> Actually, I don't see fold/unfold option at all for markdown comments. Sigh! Let me see what's going on.

This turned out to be the result of making the feature standard feature. Without that patch, the markdown folds/unfolds well. 

Anyway, I am unable to see the problem. @stephan-herrmann Can you please check now?

**Comment by stephan-herrmann** [comment]
> This turned out to be the result of making the feature standard feature. Without that patch, the markdown folds/unfolds well.

Wow, for preview features, `isSupported()` only checks that flag, ignoring the compliance / source levels. As a standard feature we check compliance, but `org.eclipse.jdt.ui.text.folding.DefaultJavaFoldingStructureProvider.fSharedScanner` uses `JavaCore`'s default compliance level, ignoring what might be configured per project. 

Sharing a scanner seems to be incompatible with version-dependent scanning :-(

**Comment by stephan-herrmann** [comment]
> Anyway, I am unable to see the problem. @stephan-herrmann Can you please check now?

When setting the workspace default compliance to 23 I can still see the same problem as before.

@jarthana Here's the difference: I normally use unix line-ends. When I change this to windows line-ends, folding is correct. Conversely, you should be able to see the bug when switching to unix line-ends. 

Does this help to find some off-by-one bug?

**Comment by jarthana** [comment]
> Wow, for preview features, `isSupported()` only checks that flag, ignoring the compliance / source levels. As a standard feature we check compliance, but `org.eclipse.jdt.ui.text.folding.DefaultJavaFoldingStructureProvider.fSharedScanner` uses `JavaCore`'s default compliance level, ignoring what might be configured per project.

Yeah, that is written based on the assumption that preview can only be enabled when the compliance is the highest. That works within the compiler but as it turns out, not all clients follow that rule.

**Comment by jarthana** [comment]
> > Anyway, I am unable to see the problem. @stephan-herrmann Can you please check now?
> 
> When setting the workspace default compliance to 23 I can still see the same problem as before.
> 
> @jarthana Here's the difference: I normally use unix line-ends. When I change this to windows line-ends, folding is correct. Conversely, you should be able to see the bug when switching to unix line-ends.
> 
> Does this help to find some off-by-one bug?

I will try this. Thanks for the tip.

**Comment by stephan-herrmann** [comment]
@jarthana do you need help here?

**Comment by stephan-herrmann** [comment]
Supposedly fixed by https://github.com/eclipse-jdt/eclipse.jdt.core/pull/3760 but I still wonder if we can create a unit test in jdt.ui?

**Comment by stephan-herrmann** [comment]
> Supposedly fixed by [eclipse-jdt/eclipse.jdt.core#3760](https://github.com/eclipse-jdt/eclipse.jdt.core/pull/3760) but I still wonder if we can create a unit test in jdt.ui?

I found no tests in jdt.ui coming close to the task.
